lua/ts-node-action/filetypes/python.lua

```diff
@@ -1,26 +1,534 @@
+-- nodes by TS.  This means that without special handling, a config of:
+-- {
+--   ["is"]  = " %s ",
+--   ["not"] = " %s "
+-- }
+-- would be padded as ` is  not `.  To avoid this, we can make them smarter by
+-- defining a padding rule for the case of when "not" is seen after "is".
+--
+-- There is also the identical case of "not in". However, "-" is both a
+-- unary and binary operator.  When it is used as a binary operator, it is
+-- a normal case.  For unary, we don't want any padding, and generally (always?)
+-- it is preceded by a named node.  When padding, we see these as
+-- prev_text=nil, so we can use that to detect the unary case with a special
+-- key "nil", to represent it.
 local padding = {
-  [","] = "%s ",
-  [":"] = "%s ",
-  ["{"] = "%s ",
-  ["}"] = " %s",
+  [","]      = "%s ",
+  [":"]      = "%s ",
+  ["{"]      = "%s",
+  ["}"]      = "%s",
+  ["for"]    = " %s ",
+  ["if"]     = " %s ",
+  ["else"]   = " %s ",
+  ["and"]    = " %s ",
+  ["or"]     = " %s ",
+  ["is"]     = " %s ",
+  ["not"]    = { [""] = " %s ", ["is"]  = "%s " },
+  ["in"]     = { [""] = " %s ", ["not"] = "%s " },
+  ["=="]     = " %s ",
+  ["!="]     = " %s ",
+  [">="]     = " %s ",
+  ["<="]     = " %s ",
+  [">"]      = " %s ",
+  ["<"]      = " %s ",
+  ["+"]      = " %s ",
+  ["-"]      = { [""] = " %s ", ["nil"] = "%s", },
+  ["*"]      = " %s ",
+  ["/"]      = " %s ",
+  ["//"]     = " %s ",
+  ["%"]      = " %s ",
+  ["**"]     = " %s ",
+  ["lambda"] = " %s ",
+  ["with"]   = " %s ",
+  ["as"]     = " %s ",
+  ["import"] = " %s ",
+  ["from"]   = "%s ",
 }
 
 local boolean_override = {
   ["True"]  = "False",
   ["False"] = "True",
 }
 
+-- @param node tsnode
+local function node_trim_whitespace(node)
+  local start_row, _, end_row, _ = node:range()
+  vim.cmd(
+    "silent! keeppatterns " .. (start_row + 1) .. "," .. (end_row + 1) ..
+    "s/\\s\\+$//g"
+  )
+end
+
+-- When inlined, these nodes must be parenthesized to avoid changing the
+-- meaning of the code and to avoid syntax errors.
+-- eg: x = lambda y: y + 1 if y else 0
+--     x = (lambda y: y + 1) if y else 0
+-- Both are valid, but the first is not equivalent to the second.
+-- Unlike the second, the first can not be expanded to:
+--    if y:
+--        x = lambda y: y + 1
+--    else:
+--        x = 0
+-- because "1 if y else 0" is inside the lambda.
+local node_types_to_parenthesize = {
+  ["conditional_expression"] = true,
+  ["boolean_operator"] = true,
+  ["lambda"] = true,
+}
+
+local function parenthesize_if_needed(node, text)
+  if node_types_to_parenthesize[node:type()] and
+    text:sub(1, 1) ~= "(" then
+    return "(" .. text .. ")"
+  end
+  return text
+end
+
+-- Recreating actions.toggle_multiline.collapse_child_nodes() here because
+-- it is not exported.  It was not possible to use helpers.node_text() on a
+-- multiline node because it will include the "\n", which is invalid for the
+-- replacement text.
```

Comment:
Do you think it would be helpful to strip the `\n` literals from the return value of `helpers.node_text()`?  I don't see the reason to keep them. It would probably make more sense to return multi-line nodes as tables, ie, something like:
```lua
function M.node_text(node)
  local text = vim.treesitter.query.get_node_text(node, vim.api.nvim_get_current_buf())
  if text:match("\n") then
    return vim.tbl_map(vim.trim, vim.split(text, "\n"))
  else
    return text
  end
end
```

Reply:
There might be some callers that are checking for the presence of `\n`, and those could be updated to just check if its a table.